examples/piggy-bank/part2/tests/tests.rs

```diff
@@ -0,0 +1,195 @@
+
+    let update = chain
+        .contract_update(
+            Signer::with_one_key(),
+            ACC_ADDR_OWNER,
+            Address::Account(ACC_ADDR_OWNER),
+            Energy::from(10000),
+            UpdateContractPayload {
+                amount:       Amount::zero(),
+                address:      initialization.contract_address,
+                receive_name: OwnedReceiveName::new_unchecked("PiggyBank.smash".to_string()),
+                message:      OwnedParameter::empty(),
+            },
+        )
+        .expect("Owner is allowed to smash intact piggy bank");
+
+    // Invoke the view function to check the state and balance.
+    let invoke = chain
+        .contract_invoke(
+            ACC_ADDR_OWNER,
+            Address::Account(ACC_ADDR_OWNER),
+            Energy::from(10000),
+            UpdateContractPayload {
+                amount:       Amount::zero(),
+                address:      initialization.contract_address,
+                receive_name: OwnedReceiveName::new_unchecked("PiggyBank.view".to_string()),
+                message:      OwnedParameter::empty(),
+            },
+        )
+        .expect("Invoking `view` should always succeed");
+
+    // Ensure the values returned by the view function are correct.
+    let (state, balance): (PiggyBankState, Amount) =
+        from_bytes(&invoke.return_value).expect("View should always return a valid result");
+    assert_eq!(state, PiggyBankState::Smashed);
+    assert_eq!(balance, Amount::zero());
+    assert_eq!(update.account_transfers().collect::<Vec<_>>(), [(
+        initialization.contract_address,
+        Amount::zero(),
+        ACC_ADDR_OWNER
+    )]);
+}
+
+#[test]
+fn test_smash_intact_not_owner() {
+    let (mut chain, initialization) = setup_chain_and_contract();
+
+    let update_err = chain
+        .contract_update(
+            Signer::with_one_key(),
+            ACC_ADDR_OTHER,
+            Address::Account(ACC_ADDR_OTHER),
+            Energy::from(10000),
+            UpdateContractPayload {
+                amount:       Amount::zero(),
+                address:      initialization.contract_address,
+                receive_name: OwnedReceiveName::new_unchecked("PiggyBank.smash".to_string()),
+                message:      OwnedParameter::empty(),
+            },
+        )
+        .expect_err("Smashing should only succeed for the owner");
+
+    let return_value =
+        update_err.return_value().expect("Contract should reject and thus return bytes");
+    let error: SmashError = from_bytes(&return_value)
+        .expect("Contract should return a `SmashError` in serialized form");
+
+    assert_eq!(error, SmashError::NotOwner, "Contract did not fail due to a NotOwner error");
+    assert_eq!(
+        chain.account_balance_available(ACC_ADDR_OTHER),
+        Some(ACC_INITIAL_BALANCE - update_err.transaction_fee),
+        "The invoker account was incorrectly charged"
+    )
+}
+
+#[test]
+fn test_smash_smashed() {
```

Comment:
It would be useful to add a sentence or two explaining what is being tested. Parts can be guessed from the name, but there are a few details that are ambiguous. This would help maintain the tests.